transports/webrtc/src/udp_mux.rs

```diff
@@ -0,0 +1,370 @@
+                    };
+
+                    let conn = match conn {
+                        // If we couldn't find the connection based on source address, see if
+                        // this is a STUN mesage and if so if we can find the connection based on ufrag.
+                        None if is_stun_message(read.filled()) => {
+                            self.conn_from_stun_message(&read.filled(), &addr)
+                        }
+                        s @ Some(_) => s,
+                        _ => None,
+                    };
+
+                    match conn {
+                        None => {
+                            if !self.new_addrs.read().contains(&addr) {
+                                match ufrag_from_stun_message(read.filled(), false) {
+                                    Ok(ufrag) => {
+                                        log::trace!(
+                                            "Notifying about new address addr={} from ufrag={}",
+                                            &addr,
+                                            ufrag
+                                        );
+                                        let mut new_addrs = self.new_addrs.write();
+                                        new_addrs.insert(addr);
+                                        return Poll::Ready(UDPMuxEvent::NewAddr(NewAddr {
+                                            addr,
+                                            ufrag,
+                                        }));
+                                    }
+                                    Err(e) => {
+                                        log::debug!(
+                                            "Unknown address addr={} (non STUN packet: {})",
+                                            &addr,
+                                            e
+                                        );
+                                    }
+                                }
+                            }
+                        }
+                        Some(conn) => {
+                            let mut packet = vec![0u8; read.filled().len()];
+                            packet.copy_from_slice(read.filled());
+                            write_packet_to_conn_from_addr(conn, packet, addr);
+                        }
+                    }
+                }
+                Err(err) if err.kind() == ErrorKind::TimedOut => {}
+                Err(err) => {
+                    log::error!("Could not read udp packet: {}", err);
+                    return Poll::Ready(UDPMuxEvent::Error(err));
+                }
+            }
+        }
+    }
+}
+
+fn write_packet_to_conn_from_addr(conn: UDPMuxConn, packet: Vec<u8>, addr: SocketAddr) {
+    tokio::spawn(async move {
+        if let Err(err) = conn.write_packet(&packet, addr).await {
+            log::error!("Failed to write packet: {} (addr={})", err, addr);
+        }
+    });
```

Comment:
Do I understand the consequences of spawning a new task here correctly?

- For each incoming packet a new task is spawned.
- Each of these tasks holds a buffer with the payload of the incoming packet.
- In case the remote is faster sending the packets than the `UdpMuxConn` is reading these packets, the number of tasks grows indefinitely.
- The tasks growing indefinitely will eventually lead to an out-of-memory kill of the process.

Reply:
I think this is correct.

Reply:
My hope would be, that we can get rid of this `spawn` with @thomaseizinger proposed design in https://github.com/melekes/rust-libp2p/pull/1.

More specifically, I would expect `UdpMuxNewAddr` to have a channel to each connection. In my eyes, there is nothing speaking against that channel having a small capacity. In case `UdpMuxNewAddr` has a packet for a connection, but the channel to that connection is full, `UdpMuxNewAddr` would no longer read from the underlying UDP socket, until that channel has capacity for the new packet. That way, we no longer break the backpressure chain.

Reply:
> UdpMuxNewAddr would no longer read from the underlying UDP socket, until that channel has capacity for the new packet.

If we don't read from the UDP socket, we may miss messages for other connections. It's a UDP socket after all. Note each `UdpMuxConn` has a buffer with max size - 4MB. If we'll be able to just write to `UdpMuxConn` buffer without `spawn`ing a thread, then buffers will act as a backpressure mechanism.

Reply:
Changed it to blocking version.

Reply:
> If we don't read from the UDP socket, we may miss messages for other connections.

Can you expand on this? I don't think we would "miss" packets. Packets should be queued by the OS and not dropped, correct?

Reply:
> Packets should be queued by the OS and not dropped, correct?

Correct. Up to a OS buffer size I believe.